ouroboros-network/src/Ouroboros/Network/PeerSelection/PublicRootPeers.hs

```diff
@@ -0,0 +1,221 @@
+import qualified Data.Set as Set
+import           Ouroboros.Network.PeerSelection.PeerAdvertise (PeerAdvertise)
+
+---------------------------------------
+-- Public root peer set representation
+--
+
+-- | Public Root Peers consist of either a set of manually configured
+-- bootstrap peers
+data PublicRootPeers peeraddr =
+  PublicRootPeers { getPublicConfigPeers :: !(Map peeraddr PeerAdvertise)
+                  , getBootstrapPeers    :: !(Set peeraddr)
+                  , getLedgerPeers       :: !(Set peeraddr)
+                  , getBigLedgerPeers    :: !(Set peeraddr)
+                  }
+  deriving (Eq, Show)
+
+invariant :: Ord peeraddr => PublicRootPeers peeraddr -> Bool
+invariant (PublicRootPeers pp bsp lp blp) =
+     -- bootstrap peers and ledger peers should not overlap
+     Set.null (Set.intersection bsp lp)
+     -- bootstrap peers and big ledger peers should not overlap
+  && Set.null (Set.intersection bsp blp)
+     -- bootstrap peers and public config peers should not overlap
+  && Set.null (Set.intersection bsp (Map.keysSet pp))
+    -- ledger and big ledger peers should not overlap
+  && Set.null (Set.intersection lp blp)
+    -- ledger peers and public config peers should not overlap
+  && Set.null (Set.intersection lp (Map.keysSet pp))
+    -- big ledger peers and public config peers should not overlap
+  && Set.null (Set.intersection blp (Map.keysSet pp))
+
+empty :: PublicRootPeers peeraddr
+empty = PublicRootPeers Map.empty Set.empty Set.empty Set.empty
+
+nullPublicConfig :: PublicRootPeers peeraddr -> Bool
+nullPublicConfig PublicRootPeers { getPublicConfigPeers } =
+  Map.null getPublicConfigPeers
+
+nullBootstrap :: PublicRootPeers peeraddr -> Bool
+nullBootstrap PublicRootPeers { getBootstrapPeers } =
+  Set.null getBootstrapPeers
+
+nullLedgerPeers :: PublicRootPeers peeraddr -> Bool
+nullLedgerPeers PublicRootPeers { getLedgerPeers } =
+  Set.null getLedgerPeers
+
+nullBigLedgerPeers :: PublicRootPeers peeraddr -> Bool
+nullBigLedgerPeers PublicRootPeers { getBigLedgerPeers } =
+  Set.null getBigLedgerPeers
+
+nullAllLedgerPeers :: PublicRootPeers peeraddr -> Bool
+nullAllLedgerPeers prp =
+  nullLedgerPeers prp && nullBigLedgerPeers prp
+
+null :: PublicRootPeers peeraddr -> Bool
+null prp = nullPublicConfig prp
+        && nullBootstrap prp
+        && nullAllLedgerPeers prp
+
+size :: PublicRootPeers peeraddr -> Int
+size (PublicRootPeers pp bsp lp blp) =
+    Map.size pp
+  + Set.size bsp
+  + Set.size lp
+  + Set.size blp
+
+member :: Ord peeraddr => peeraddr -> PublicRootPeers peeraddr -> Bool
+member p (PublicRootPeers pp bsp lp blp) =
+     Map.member p pp
+  || Set.member p bsp
+  || Set.member p lp
+  || Set.member p blp
+
+merge :: Ord peeraddr
+      => PublicRootPeers peeraddr
+      -> PublicRootPeers peeraddr
+      -> PublicRootPeers peeraddr
+merge (PublicRootPeers pp bsp lp blp) (PublicRootPeers pp' bsp' lp' blp') =
+  fromMapAndSet (pp <> pp') (bsp <> bsp') (lp <> lp') (blp <> blp')
```

Comment:
You could trun this into a monoid instance.

Reply:
unfortunately `merge` requires an `Ord` instance so it is not that easy

Reply:
That's surprising, `Map` has a `Monoid` [instance](https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Strict.html#t:Map); this should work for us too:
```hs
instance Ord peeraddr => Monoid (PublicRootPeers peeraddr) where
```
It shouldn't require adding `Ord peeraddr` instance at call site, since `merge` already requires it as well.

Reply:
That's true, I was being dumb